pkg/registry/query.go

```diff
@@ -0,0 +1,320 @@
+				}
+				bundles = append(bundles, apiBundle)
+			}
+		}
+	}
+	return bundles, nil
+}
+
+func (q Querier) GetPackage(_ context.Context, name string) (*PackageManifest, error) {
+	pkg, ok := q.pkgs[name]
+	if !ok {
+		return nil, fmt.Errorf("package %q not found", name)
+	}
+
+	var channels []PackageChannel
+	for _, ch := range pkg.Channels {
+		head, err := ch.Head()
+		if err != nil {
+			return nil, fmt.Errorf("package %q, channel %q has invalid head: %v", name, ch.Name, err)
+		}
+		channels = append(channels, PackageChannel{
+			Name:           ch.Name,
+			CurrentCSVName: head.Name,
+		})
+	}
+	return &PackageManifest{
+		PackageName:        pkg.Name,
+		Channels:           channels,
+		DefaultChannelName: pkg.DefaultChannel.Name,
+	}, nil
+}
+
+func (q Querier) GetBundle(_ context.Context, pkgName, channelName, csvName string) (*api.Bundle, error) {
+	pkg, ok := q.pkgs[pkgName]
+	if !ok {
+		return nil, fmt.Errorf("package %q not found", pkgName)
+	}
+	ch, ok := pkg.Channels[channelName]
+	if !ok {
+		return nil, fmt.Errorf("package %q, channel %q not found", pkgName, channelName)
+	}
+	b, ok := ch.Bundles[csvName]
+	if !ok {
+		return nil, fmt.Errorf("package %q, channel %q, bundle %q not found", pkgName, channelName, csvName)
+	}
+	apiBundle, err := api.ConvertModelBundleToAPIBundle(*b)
+	if err != nil {
+		return nil, fmt.Errorf("convert bundle %q: %v", b.Name, err)
+	}
+
+	// unset Replaces and Skips (sqlite query does not populate these fields)
+	apiBundle.Replaces = ""
+	apiBundle.Skips = nil
+	return apiBundle, nil
+}
+
+func (q Querier) GetBundleForChannel(_ context.Context, pkgName string, channelName string) (*api.Bundle, error) {
+	pkg, ok := q.pkgs[pkgName]
+	if !ok {
+		return nil, fmt.Errorf("package %q not found", pkgName)
+	}
+	ch, ok := pkg.Channels[channelName]
+	if !ok {
+		return nil, fmt.Errorf("package %q, channel %q not found", pkgName, channelName)
+	}
+	head, err := ch.Head()
+	if err != nil {
+		return nil, fmt.Errorf("package %q, channel %q has invalid head: %v", pkgName, channelName, err)
+	}
+	apiBundle, err := api.ConvertModelBundleToAPIBundle(*head)
+	if err != nil {
+		return nil, fmt.Errorf("convert bundle %q: %v", head.Name, err)
+	}
+
+	// unset Replaces and Skips (sqlite query does not populate these fields)
+	apiBundle.Replaces = ""
+	apiBundle.Skips = nil
+	return apiBundle, nil
+}
+
+func (q Querier) GetChannelEntriesThatReplace(_ context.Context, name string) ([]*ChannelEntry, error) {
+	var entries []*ChannelEntry
+
+	for _, pkg := range q.pkgs {
+		for _, ch := range pkg.Channels {
+			for _, b := range ch.Bundles {
+				entries = append(entries, channelEntriesThatReplace(*b, name)...)
+			}
+		}
+	}
+	if len(entries) == 0 {
+		return nil, fmt.Errorf("no channel entries found that replace %s", name)
+	}
+	return entries, nil
+}
+
+// TODO(joelanford): What if multiple bundles replace this one?
+//    Don't worry about this. Not used anymore.
```

Comment:
This one appears to still used by catalog-operator:

https://github.com/operator-framework/operator-lifecycle-manager/blob/master/pkg/controller/operators/catalog/operator.go#L1011
https://github.com/operator-framework/operator-lifecycle-manager/blob/d9ff24161ab4f7d4d323afb662a56f453791329c/pkg/controller/registry/resolver/querier.go#L143
https://github.com/operator-framework/operator-registry/blob/a1fc9d0f1d2e04c14229801605ed727cd80e956e/pkg/client/client.go#L74

Reply:
I'll double check, but I think there's some dead code in OLM, and this might be one of those cases.

Reply:
Oh now I see you already did the leg work. Thanks!

In that case, I'll cross-reference with the sqlite implementation of this function and see if its deterministic (and if so, I'll use the same algorithm).

Reply:
The sqlite implementation looks non-deterministic to me since there's no ORDER BY in the query.

The model implementation in this PR is ALSO non-deterministic because it iterates over a map and chooses the first bundle it finds that meets the criteria. Since map iteration is non-deterministic, the returned bundle is non-deterministic.

I'll remove this TODO comment, and add some comments in the code that call this out.

Reply:
Done!